pageserver/src/tenant/timeline.rs

```diff
@@ -4062,17 +4121,11 @@ impl Timeline {
 
     /// Return true if the value changed
     ///
-    /// This function must only be used from the layer flush task, and may not be called concurrently.
+    /// This function must only be used from the layer flush task.
     fn set_disk_consistent_lsn(&self, new_value: Lsn) -> bool {
-        // We do a simple load/store cycle: that's why this function isn't safe for concurrent use.
-        let old_value = self.disk_consistent_lsn.load();
-        if new_value != old_value {
-            assert!(new_value >= old_value);
-            self.disk_consistent_lsn.store(new_value);
-            true
-        } else {
-            false
-        }
+        let old_value = self.disk_consistent_lsn.fetch_max(new_value);
+        assert!(new_value >= old_value);
+        new_value != old_value
     }
```

Comment:
This code is asserting _after_ the damage is done. Why not keep the current code?

Reply:
fetch_max does not modify the value if the stored value is larger than the given value, so no damage did get to happen.

I did split this out to a separate PR. i thought it is a footgun to have this non-atomic lockless code lying around which has 2 callsites.

Reply:
I've already split this to another PR #8311.